src/loader.c

```diff
@@ -164,13 +169,42 @@ static bool set_auth_dbname(PgDatabase *db, const char *new_auth_dbname)
 
 static bool set_autodb(const char *connstr)
 {
-	char *tmp = strdup(connstr);
+    char *p, *key, *val;
+    char *tmp_connstr = NULL;
+	char *tmp = NULL;
 	char *old = cf_autodb_connstr;
 
-	if (!tmp) {
-		log_error("no mem to change autodb_connstr");
-		return false;
-	}
+    /* Validating constraints for database connection string */
+    tmp_connstr = strdup(connstr);
+    if (!tmp_connstr) {
+        log_error("no mem to change autodb_connstr");
+        return false;
+    }
+
+    p = tmp_connstr;
+    while (*p) {
```

Comment:
This loop is not required. The "cf_autodb_connstr" will be given to parse_database function, which will detect that "auth_dbname" is set incorrectly.

Reply:
I agree, I will move this check inside the `parse_database` function.

Reply:
I do not think there is a need to add any other check than one you've already added to set_authdb function, which makes pgbouncer comparision. That should be sufficient.

Reply:
Referring to the [line](https://github.com/pgbouncer/pgbouncer/blob/8d0ba3bd8e7e0afe10ebf065e52baa3a14666bbf/src/loader.c#L260) it seems that if we move checks within the `parse_database` function we should partially change parsing logic. 

We could implement a similar as below:
```
bool parse_database(void *base, const char *name, const char *connstr)
{
        /* ...................... */
	bool is_autodb = false;
        /* ...................... */
        if (strcmp(name, "*") == 0) {
                is_autodb = true
                if (!set_autodb(connstr)){
                    return false   
                }
	}
        
        /* ...................... */
	while (*p) {
            else if (strcmp("auth_dbname", key) == 0) {
                        /* Added reserved auth_dbname check */
                        if (check_reserved_database(key)}
                            auth_dbname = val;
                        } else {
                            return false;
                        }
            }
       }
       if (is_autodb) {
           // We do not need to execute further code and allocate memory 
           // for autodb using the `add_database` function. Instead, we have 
           // passed through the all connstr definition and returns true (all is valid)
           return true;
       }
        /* ...................... */
}
```

The main point is - autodb is not a real DB in the pgbouncer terms during the config parsing stage so that we store this value into `cf_autodb_connstr` and do not create a real DB, instead we are creating it when autodb connection is matched. According to that, we can make some checks after the loop (is it autodb or not) in order to avoid memory allocation using the `add_database` function.

I would be glad to see any comments about this concern

Reply:
I see your point that this would prevent "*" being **set** incorrectly. However, I do not see any benefit of performing this additional check with extra complexity. Because, connection attempts landing on the autodb will get rejected due to incorrect configuration anyway once the registered connection string is provided to parse_database.

Reply:
@goksgie you took the point across correctly. But in my point of view, it is better when in the config applying stage we would find any misconfiguration issues. Anyway, I will remove this additional check.

Reply:
I closed this thread because I reverted the changes